core/lib/Hooks.re

```diff
@@ -0,0 +1,266 @@
+type hook('a) = ..;
+
+module Slots =
+  Slots.Make({
+    type t('a) = hook('a);
+  });
+
+type t('a, 'b) = {
+  slots: Slots.t('a, 'b),
+  onSlotsDidChange: unit => unit,
+};
+
+type empty = t(unit, unit);
+
+type hooks('slots, 'nextSlots) = t('slots, 'nextSlots);
+
+let create = (~onSlotsDidChange) => {
+  slots: Slots.create(),
+  onSlotsDidChange,
+};
+
+module State = {
```

Comment:
As a fun idea,  in `reason-reactify` we used to implement `useState` as a special case of `useReducer` which allowed to reduce some of the boilerplate. Not sure if it'd be interesting in this case :) 

https://github.com/revery-ui/reason-reactify/blob/38e59a7726146bdb63e4b53bb3068bab6132aa95/lib/Reactify.re#L565-L577

Reply:
they were implemented differently on purpose. My thought process is that the operation in the reducer might take relatively long to compute and actions are very often fired during interactions. Enqueuing an update takes constant time. Also, with this approach it'll be relatively easy to implement `UpdateWithSideEffects`. That said I'm not sure about this myself 😕